### PR TITLE
Jetpack Cloud: Refactor landing logic to avoid extra redirects

### PR DESCRIPTION
**Pull request by mattgawarecki**
When a visitor to Jetpack Cloud clicks on the "Manage Sites" link at the top of the page (e.g., if they're currently on the Licensing page), they see the page flicker and fully reload, even though with React this is unnecessary. This PR fixes the issue by moving the landing page redirection logic into a shared file and only reloading the page when site data is missing and must be fetched.

Resolves `1202858161751496-as-1202238386365801`.

## Proposed Changes

* Move the `landing.tsx` component to `landing/main.tsx`.
* Create a new `selectors.ts` file containing selectors for determining the correct landing page for any Jetpack Cloud visitor.
* Use the new selectors in the app's client-side router (in `index.js`), to avoid loading the `<Landing />` component if at all possible. (The component will still load if required site data is missing and must be fetched.)
* Add unit test coverage for the new `getLandingPath` and `isSiteEligibleForJetpackCloud` selectors.

## Known Issues

* Self-hosted sites with Jetpack Social should land on the Social page. This will be resolved in `1202858161751496-as-1203937561792507`.

### Extra page loading/flickering (verify bug fix)

0. Log into Jetpack Cloud with a WordPress.com account that has access to the Licensing page. (A12s: for help with this, contact @Automattic/jetpack-genesis)
4. Click the "Licensing" link to visit the Licensing page.
5. Click the "Manage Sites" link.
6. Verify the page does not flicker and fully reload.
7. Click the "Licensing" link again.
8. Verify the page does not flicker and fully reload.

### Landing page (regression testing)

_For the purposes of all landing page tests, reference the following mapping of active subscriptions to the expected landing page:_

#### Ineligible sites

| Site type | Landing page |
| --------- | ------------ |
| WordPress.com Simple | `/landing` |
| Atomic | `/landing` |
| Multi-site | `/landing` |

_(* Note: unit tests in this PR assert that `/` is the correct landing page. This is technically true, but we redirect immediately to `/landing` from there.)_

#### Self-hosted single Jetpack sites

| Product(s) | Landing page |
| ---------- | ------------ |
| No product (i.e., Jetpack Free) | `/backup/<site>` |
| Backup alone | `/backup/<site>` |
| Backup + any other product(s) | `/backup/<site>` |
| Scan alone | `/scan/<site>` |
| Scan + any other product (excl. Backup) | `/scan/<site>` |
| Search alone | `/jetpack-search/<site>` |
| Search + any other product (excl. Backup and Scan) | `/jetpack-search/<site>` |

#### Landing on the correct page: selected site

1. Select or create a new site with each of the combinations from the above table; then, visit the landing URL for this site in your testing environment (e.g., `http://jetpack.cloud.localhost:3000/landing/mysite.example`). Verify you're immediately sent to the corresponding page.

#### Landing on the correct page: primary site

0. Using the above table, determine which is the appropriate landing page for the site you've marked in your WordPress.com account settings (https://wordpress.com/me/account) as your primary site. If necessary, change your primary site to match the test criteria.
1. Visit the root URL of Jetpack Cloud in your test environment.
2. Using the table above, verify you're automatically sent to the correct page.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

**Comment by manzoorwanijk** [comment]
Is this ready for review?

**Comment by mattgawarecki** [comment]
> Is this ready for review?

Technically yes, but I hadn't yet finished the full set of testing instructions, so I left it in draft state 😅

File: client/jetpack-cloud/index.js [hunk 1]
```diff
@@ -52,8 +55,31 @@ const redirectToPrimarySiteLanding = ( context, next ) => {
...
 	debug( 'controller: landingController', context );
-	context.primary = <Landing />;
-	next();
+
+	const state = context.store.getState();
+	const selectedSiteId = getSelectedSiteId( state );
+
+	// This path requires a selected site ID;
+	// if we don't have one, go to the main landing page
+	if ( ! Number.isFinite( selectedSiteId ) ) {
```
Comment (TheSteveK) [inline client/jetpack-cloud/index.js]: Not a review: I hadn't seen isFinite before.  Are there situations where it's preferable to looking for a null instead?

Comment (mattgawarecki) [reply client/jetpack-cloud/index.js]: > Are there situations where it's preferable to looking for a null instead?

_Hypothetically_, if you're trying to determine whether a number input is a "normal" number, `Number.isFinite` is the function to use. In this particular case, I don't think it makes much of a difference other than to signal we're looking at a variable that should contain a numeric value of some kind. The possibilities shape up like so (notice how the outputs for `0`, `Infinity`, and `'abc'` change):

`Number.isFinite`:

| Input     | Output |
| --------- | ------ |
| 4         | true   |
| 0         | true   |
| NaN       | false  |
| Infinity  | false  |
| null      | false  |
| undefined | false  |
| 'abc'     | false  |

Truthy check (e.g., `if ( x )`):

| Input     | Output |
| --------- | ------ |
| 4         | true   |
| 0         | false  |
| NaN       | false  |
| Infinity  | true   |
| null      | false  |
| undefined | false  |
| 'abc'     | true   |

**Review by elliottprogrammer — APPROVED** [review]
Performed all tests.  Everything works exactly as described. No flickering switching between "Manage sites" and "Licensing" links. All code looks very good to me. I don't see any issues.
Unit tests look good too, and all passing.
LGTM! 👍

**Comment by elliottprogrammer** [comment]
Although I'm pretty sure this is unrelated to this PR, when I tested a site with 1 purchased subscription for the "Site Search" product (no other products on the site), the landing page redirect worked great!, just as described it landed on `/jetpack-search/<site>`.  However, for some reason the ` /jetpack-search/<site>` page is saying that Search is _Disabled_, even though in the Jetpack plugin Search page it is fully enabled, and in the Jetpack plugin Settings --> Performance page it is fully enabled the same.

Here is a video showing the issue more clearly.

https://user-images.githubusercontent.com/11078128/219792898-94ef4e12-9da0-4234-85af-c9858c90d428.mp4

.
Ahhh, ok yes, I tested this on `trunk` branch and the issue persists. So it's **Not** related to this PR. It's probably just an overlooked issue that occurred at some point and a maintenance task for it should be created.
I've created a maintenance ticket on the M-Origin Asana board. --> 1202858161751496-as-1203986291851623/f

Carry on..  This PR is good. 😉

**Pull request merged**